### PR TITLE
KAFKA-7061: Enhanced log compaction

### PR DESCRIPTION
**Pull request by blaghed**
Enhance log compaction to support more than just offset comparison, so the insertion order isn't dictating which records to keep.

Default behavior is kept as it was, with the enhanced approached having to be purposely activated.
The enhanced compaction is done either via the record timestamp, by settings the new configuration as "timestamp" or via the record headers by setting this configuration to anything other than the default "offset" or the reserved "timestamp".

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by blaghed** [comment]
This picks up where https://github.com/apache/kafka/pull/4822 left off

**Comment by blaghed** [comment]
retest this please

**Comment by blaghed** [comment]
retest this please

File: core/src/main/scala/kafka/log/CleanerCache.scala [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+/**
+ * A hash table used for deduplicating the log.
+ * This hash table uses a cryptographically secure hash of the key as a proxy
+ * for the key for comparisons and to save space on object overhead.
+ * Collisions are resolved by probing.
+ * This hash table does not support deletes.
+ *
+ * @param memory        The amount of memory this map can use
+ * @param hashAlgorithm The hash algorithm instance to use: MD2, MD5, SHA-1, SHA-256, SHA-384, SHA-512
```
Comment (BertusGreeff) [inline core/src/main/scala/kafka/log/CleanerCache.scala:97]: I don't know the Kafka code base, so please take this in context, but my feedback here is that using cryptographic hashes for dictionary hashing is not a good approach.  Its too slow.  Just use Java's native dictionary implementation.

Comment (blaghed) [reply core/src/main/scala/kafka/log/CleanerCache.scala:97]: Hi @BertusGreeff . Please refer to the original code:
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/OffsetMap.scala

You can, of course, file a change to address that.

**Pull request closed without merging**